### ISSUE TITLE
Allow basic math/... expressions in the const syntax

### ISSUE THREAD

**Issue by Janiczek** [gleam-lang/gleam#1488]
Context: [Discord discussion](https://discord.com/channels/768594524158427167/811254431051808778/939533231936659527)

<img width="834" alt="Screenshot 2022-02-05 at 16 00 06" src="https://user-images.githubusercontent.com/149425/152647137-ae94c579-e1d1-4298-b274-48a4f05a1a18.png">

**Comment by lpil** [comment]
Thanks!

**Comment by GearsDatapacks** [comment]
Hi, I'm looking into implementing this. Should this allow all binary operators in constant expressions or just a subset? And should the output of those operations be computed at compile-time (`1 + 2` compiles to `3`), or should it just emit a binary operation in the generated code (Like string concatenation is done currently)
I imagine the latter will be simpler, although I'm not sure if any limitations apply on the erlang side of things

**Comment by lpil** [comment]
It needs to be at compile time as otherwise they are not constants and have a runtime cost to using.

**Comment by GearsDatapacks** [comment]
It does on javascript. I can change that now though. And should all binary operations be allowed?

**Comment by lpil** [comment]
JS is a scripting language so it's computed once at load time and there's no cost per-use. We can leave it as is.

**Comment by GearsDatapacks** [comment]
So I'm looking into implementing this once again, and I would like to know people's opinions on how it should be implemented.
My initial thought is (at least for integers), compute any constant-time operations using a `BigInt` type. This prevents any overflow errors/unexpected behaviour due to the rust implementation. Once calculated, it would be converted to a literal, and then whatever target-specific semantics would be applied.
For floats it's a little trickier, since target-specific semantics can be much more finicky and nuanced. I don't think there's an easy way to perfectly replicate it at compile-time, without completely reimplementing the exact semantics per-target (which we probably don't want to have to do). We could do the same thing and compute them with arbitrary precision floats; would that cause confusing bugs? I'm not too sure.

**Comment by seleniumforest** [comment]
Bump. Would be useful to work with time ranges and unix dates. 

```gleam
const minute = 60
const hour = minute * 60
const day = hour * 24

fn foo(){
  let now = current_unix_time()

  //now it's easy to work with unix date, make plus 3 days 12 hours 25 minutes
  let some_date = now + (day * 3) + (hour * 12) + (minute * 25)
}

fn current_unix_time() -> Int {
  let #(now, _) =
    timestamp.system_time()
    |> timestamp.to_unix_seconds_and_nanoseconds
  now
}
```

**Comment by giacomocavalieri** [comment]
> So I'm looking into implementing this once again, and I would like to know people's opinions on how it should be implemented.
My initial thought is (at least for integers), compute any constant-time operations using a BigInt type. This prevents any overflow errors/unexpected behaviour due to the rust implementation. Once calculated, it would be converted to a literal, and then whatever target-specific semantics would be applied.

Sounds good to me, why not start with just integers and see how that works out? Or would any implementation need to support both integers and floats?

**Comment by GearsDatapacks** [comment]
Nope, it's definitely possible to just do `Int`s. I'm happy to go ahead with this if Louis agrees with this approach

**Comment by lpil** [comment]
Sounds good to me! It's easier to reason about the semantics of ints also.

**Comment by GearsDatapacks** [comment]
Cool, I'll start work on this then! I do have working code from over a year ago, but I don't think it's worth digging up because the work in resolving all the merge conflicts since then is probably more than just re-implementing it at this point.
